Fix android prebuilt toolchains

Cloudef opened the pull request:
###### Motivation for this change

The current android toolchains are horribly broken. They either don't build code at all or produce broken executables.
I picked up https://github.com/NixOS/nixpkgs/pull/113122 work and fixed things up more. Also updated the ndk / android package lists.

You can test basic operation with the following nix file:
```nix
{ pkgs ? import (fetchTarball "https://github.com/Cloudef/nixpkgs/archive/android-prebuilt.tar.gz") {} }:

# hello.c
# #include <stdio.h>
# int main() {
#    printf("hello world\n");
#    return 0;
# }

pkgs.pkgsCross.armv7a-android-prebuilt.stdenv.mkDerivation {
   name = "hello";
   src = ./.;
   nativeBuildInputs = [ pkgs.which ];
   buildPhase = "echo $CC; which $CC; $CC --version; $CC hello.c -o hello";
   installPhase = "install -Dm755 hello -t $out";
}
```

It should give you a dynamically linked binary:
```
result/hello: ELF 32-bit LSB pie executable, ARM, EABI5 version 1 (SYSV), dynamically linked, interpreter /system/bin/linker, not stripped
```
Which will run on real device:
```
:/data/data/org.galexander.sshd/files $ ./hello
hello world
```

Also tested on another project:
```
:/data/data/org.galexander.sshd/files $ ./box86
Box86 with Dynarec v0.2.5 da6b6bb built on Feb 18 2022 04:10:17


This is Box86, the Linux x86 emulator with a twist

Usage is box86 [options] path/to/software [args]
to launch x86 software
```

Also wine compiled for i686-android running through the above emulator inside phone:
```
data/user/0/org.galexander.sshd/files/box86 wine                                                                                              <
Box86 with Dynarec v0.2.5 da6b6bb built on Feb 18 2022 04:10:17
Warning, PreInit Array (size=8) present and ignored!
Usage: wine PROGRAM [ARGUMENTS...]   Run the specified program
       wine --help                   Display this help and exit
       wine --version                Output version information and exit
```

###### Things done

- Built on platform(s)
  - [x] x86_64-linux
  - [ ] aarch64-linux
  - [ ] x86_64-darwin
  - [ ] aarch64-darwin
- [x] For non-Linux: Is `sandbox = true` set in `nix.conf`? (See [Nix manual](https://nixos.org/manual/nix/stable/command-ref/conf-file.html))
- [ ] Tested, as applicable:
  - [NixOS test(s)](https://nixos.org/manual/nixos/unstable/index.html#sec-nixos-tests) (look inside [nixos/tests](https://github.com/NixOS/nixpkgs/blob/master/nixos/tests))
  - and/or [package tests](https://nixos.org/manual/nixpkgs/unstable/#sec-package-tests)
  - or, for functions and "core" functionality, tests in [lib/tests](https://github.com/NixOS/nixpkgs/blob/master/lib/tests) or [pkgs/test](https://github.com/NixOS/nixpkgs/blob/master/pkgs/test)
  - made sure NixOS tests are [linked](https://nixos.org/manual/nixpkgs/unstable/#ssec-nixos-tests-linking) to the relevant packages
- [ ] Tested compilation of all packages that depend on this change using `nix-shell -p nixpkgs-review --run "nixpkgs-review rev HEAD"`. Note: all changes have to be committed, also see [nixpkgs-review usage](https://github.com/Mic92/nixpkgs-review#usage)
- [x] Tested basic functionality of all binary files (usually in `./result/bin/`)
- [22.05 Release Notes (or backporting 21.11 Release notes)](https://github.com/NixOS/nixpkgs/blob/master/CONTRIBUTING.md#generating-2205-release-notes)
  - [ ] (Package updates) Added a release notes entry if the change is major or breaking
  - [ ] (Module updates) Added a release notes entry if the change is significant
  - [ ] (Module addition) Added a release notes entry if adding a new NixOS module
  - [ ] (Release notes changes) Ran `nixos/doc/manual/md-to-db.sh` to update generated release notes
- [ ] Fits [CONTRIBUTING.md](https://github.com/NixOS/nixpkgs/blob/master/CONTRIBUTING.md).

Cloudef committed:
update android targets to recommended ones

SuperSandro2000 commented on pkgs/development/androidndk-pkgs/androidndk-pkgs.nix:
```diff
@@ -21,62 +24,121 @@ let
   targetInfo = ndkInfoFun stdenv.targetPlatform;
 
-  prefix = lib.optionalString (stdenv.targetPlatform != stdenv.hostPlatform) (stdenv.targetPlatform.config + "-");
+  sdkVer = stdenv.targetPlatform.sdkVer;
```
```suggestion
  inherited (stdenv.targetPlatform) sdkVer;
```

SuperSandro2000 commented on pkgs/development/androidndk-pkgs/androidndk-pkgs.nix:
```diff
@@ -21,62 +24,121 @@ let
+      for f in $out/bin/${targetInfo.triple}-*; do
+        ln -s $f ''${f/${targetInfo.triple}-/${targetConfig}-}
+      done
+      for f in `find $out/toolchain -type d -name ${targetInfo.triple}`; do
```
```suggestion
      for f in $(find $out/toolchain -type d -name ${targetInfo.triple}); do
```
backticks are legacy.

SuperSandro2000 commented on pkgs/development/androidndk-pkgs/androidndk-pkgs.nix:
```diff
@@ -21,62 +24,121 @@ let
+      # LEGACY: ld doesn't properly include transitive library dependencies.
+      #         Let's use gold instead
+      rm -f $out/bin/${targetConfig}-ld
+      if [ -f  $out/bin/${targetConfig}-ld.gold ]; then
```
```suggestion
      if [[ -f  $out/bin/${targetConfig}-ld.gold ]]; then
```
[[ prevents issues around quoting.

SuperSandro2000 commented on pkgs/development/androidndk-pkgs/androidndk-pkgs.nix:
```diff
@@ -21,62 +24,121 @@ let
     inherit (androidndk) version;
     isClang = true; # clang based cc, but bintools ld
-    nativeBuildInputs = [ makeWrapper ];
+    nativeBuildInputs = [ makeWrapper python autoPatchelfHook ];
```
Can we use python3?

SuperSandro2000 replied:
If it is unused sure. Less python 2 is always better.

matthewbauer commented:
@ofborg build pkgsCross.aarch64-android-prebuilt.hello

matthewbauer reviewed (approved):
approved pending @SuperSandro2000 's requests

Cloudef commented:
Btw, we can test that if simple binaries like hello works with this: https://github.com/Cloudef/android2gnulinux

Mindavi commented:
@Cloudef are you willing to update based on review comments?

Cloudef commented:
@Mindavi I'll see if I can work on this tomorrow, I will need this eventually.

Cloudef committed:
androidenv: update packages

Cloudef committed:
canExecute: check for android

Cloudef commented:
Updated the PR, updated the repo.json, and also fixed issue where -pie linker flag is needed for executables. So I used a slight hack to achieve this:
https://github.com/NixOS/nixpkgs/blob/498d7618d8c25e9bc9d71c8a0b45cd6fbd06d51b/pkgs/development/androidndk-pkgs/androidndk-pkgs.nix#L134

Cloudef committed:
androidenv: fix android cross-compilers

These were completely broken. Even if they managed to output a binary,
it wouldn't actually run on the actual device.

Cloudef committed:
androidndk: remove legacy ndks

allows us to remove python2 dependencies and lot more

Cloudef commented:
https://github.com/NixOS/nixpkgs/commit/b5c2b76cf4e1888b231bd011c4e699ec937b23ac caused merge conflict .. rebased

Mindavi commented:
Think this should be good, but I really don't know too much about it. If I or anyone else is willing to build the example I think that's all we can do. I think it's kinda sad if we let this go stale

Mindavi commented on lib/systems/default.nix:
```diff
@@ -36,6 +36,7 @@ rec {
       config = parse.tripleFromSystem final.parsed;
       # Determine whether we can execute binaries built for the provided platform.
       canExecute = platform:
+        final.isAndroid == platform.isAndroid &&
```
This doesn't make sense to me.

Cloudef replied:
Do you have better idea? It would be better for this to check libc too btw.

Mindavi replied:
Musl and glibc can both be executed on the same machine.

What if this is about linux.musl vs linux.glibc? Then this would evaluate to false, right? It should evaluate to true in that case, that's my main point.

Cloudef replied:
This is about android bionic libc binary trying to be executed on GNU/Linux host, you can't do that without having android linker and bionic libc on host.

Mindavi replied:
I understand. But isCompatible should return true when it's 64-bit to 32-bit 'normal' linux

Cloudef replied:
In this case it will

Mindavi replied:
Sorry, I guess I am still sleepy :) you're right
